service/apigateway/unmarshal.go

```diff
@@ -0,0 +1,509 @@
+			val, err := decoder.Token()
+			if err != nil {
+				return awserr.New(aws.ErrCodeSerialization,
+					fmt.Sprintf("failed to decode response body with invalid JSON. Here's a snapshot: %s",
+						getSnapshot(ringBuffer)), err)
+			}
+			if v, ok := val.(float64); ok {
+				time := time.Unix(int64(v), 0).UTC()
+				output.LastUpdatedDate = &time
+			} else {
+				return awserr.New(aws.ErrCodeSerialization,
+					fmt.Sprintf("Expected LastUpdatedDate to be of type timestamp. Here's a snapshot: %s",
+						getSnapshot(ringBuffer)), err)
+			}
+		}
+
+	}
+	// end of the json response body
+	endToken, err := decoder.Token()
+	if err != nil {
+		return err
+	}
+	if t, ok := endToken.(json.Delim); !ok || t.String() != "}" {
+		return awserr.New(aws.ErrCodeSerialization,
+			fmt.Sprintf("failed to decode response body with invalid JSON, expected `}` as end Token. "+
+				"Here's a snapshot: %s",
+				getSnapshot(ringBuffer)), err)
+	}
+
+	return nil
+}
+
+func unmarshalStageKeysListShapeAWSJSON(list *[]string, decoder *json.Decoder, ringBuffer *sdkio.RingBuffer) error {
+	// start of the list
+	startToken, err := decoder.Token()
+
+	if err == io.EOF {
+		// "Empty List"
+		return nil
+	}
+	if err != nil {
+		return awserr.New(aws.ErrCodeSerialization,
+			fmt.Sprintf("failed to decode response body with invalid JSON. Here's a snapshot: %s",
+				getSnapshot(ringBuffer)), err)
+	}
+	if t, ok := startToken.(json.Delim); !ok || t.String() != "[" {
+		return awserr.New(aws.ErrCodeSerialization,
+			fmt.Sprintf("Expected json delimiter at the start of list, found %v instead. "+
+				"Here's a snapshot: %s", t, getSnapshot(ringBuffer)), err)
+	}
+
+	for decoder.More() {
+		token, err := decoder.Token()
+		if err != nil {
+			return awserr.New(aws.ErrCodeSerialization,
+				fmt.Sprintf("failed to decode response body with invalid JSON. Here's a snapshot: %s",
+					getSnapshot(ringBuffer)), err)
+		}
+		// based on struct Stage
+		if v, ok := token.(string); ok {
+			val := v
+			*list = append(*list, val)
+		} else {
+			return awserr.New(aws.ErrCodeSerialization,
+				fmt.Sprintf("Expected stageKey to be of type string. Here's a snapshot: %s",
+					getSnapshot(ringBuffer)), err)
+		}
+	}
+
+	// end of the list
+	endToken, err := decoder.Token()
+	if err != nil {
+		return awserr.New(aws.ErrCodeSerialization,
+			fmt.Sprintf("failed to decode response body with invalid JSON. Here's a snapshot: %s",
+				getSnapshot(ringBuffer)), err)
+	}
+	if t, ok := endToken.(json.Delim); !ok || t.String() != "]" {
+		return awserr.New(aws.ErrCodeSerialization,
+			fmt.Sprintf("failed to decode response body with invalid JSON, expected `]` as end Token. "+
+				"Here's a snapshot: %s",
+				getSnapshot(ringBuffer)), err)
+	}
+	return nil
+}
+
+func unmarshalTagsMapShapeAWSJSON(output *map[string]string, decoder *json.Decoder, ringBuffer *sdkio.RingBuffer) error {
+	startToken, err := decoder.Token()
+	if err == io.EOF {
+		// "Empty List"
+		return nil
+	}
```

Comment:
Does this actually indicate the end of the map in this case? Or does this actually indicate the end of the entire IO stream?